OpenAI-DotNet/BaseEndPoint.cs

```diff
@@ -1,20 +1,44 @@
-﻿namespace OpenAI
+﻿using System;
+
+namespace OpenAI
 {
     public abstract class BaseEndPoint
     {
-        protected readonly OpenAIClient Api;
+        protected readonly BaseOpenAIClient Api;
+
+        /// <summary>
+        /// Delegate method for creating Url's dynamically for a given engine deployment type.
+        /// </summary>
+        /// <param name="deployment"></param>
+        /// <returns></returns>
+        public delegate string UrlGenerator(Engine deployment);
+
+        /// <summary>
+        /// The Url Generator for each endpoint, changes based on whether the service is OAI or AOAI
+        /// </summary>
+        protected readonly UrlGenerator Generator;
 
         /// <summary>
         /// Constructor of the api endpoint.
         /// Rather than instantiating this yourself, access it through an instance of <see cref="OpenAIClient"/>.
         /// </summary>
-        internal BaseEndPoint(OpenAIClient api) => Api = api;
+        internal BaseEndPoint(BaseOpenAIClient api, UrlGenerator generator)
```

Comment:
Not really a fan of this url generator delegate. It'll also add a bit of additional cost when executing.

Reply:
What are some other ways we can solve this problem?

Reply:
To make sure I understand this, is the concern over the formatting/overhead of generating the string templates on every request, or the use of a delegate function being passed into the BaseEndPoint?

I'm not a big fan of delegates myself, normally I would add another class/interface to handle the URL generation, although the delegates simplified some of the flow for this.

Quick solutions I can think of based on the current design.

- Push the URL generation into the interfaces (e.g. ICompletion interface), have the OpenAIClient/AzureOpenAIClient implement the methods in the base class.
- Require the BaseEndPoint subclasses to have separate subclasses for both OpenAI and AzureOpenAI, although this would double the number of endpoints (inheritance vs composition).
- Rewrite the delegate to IUrlGenerator, with an AzureOpenAIUrlGenerator, and a OpenAIUrlGenerator that handles the generation. This is fairly similar to the first approach, although all the logic for handling the generation of the Url's would be in a separate class.

I'm open to other opinions/suggestions.

Reply:
> the use of a delegate function being passed into the BaseEndPoint

This

Reply:
The interface solution sounds interesting

Reply:
I could see either 1 or 3 be solid solutions